Add API for enabling/disabling tagging (part of #608).

sebright opened the pull request:
This commit adds an enum to represent the tagging state (TaggingState) and adds
getters and setters for the state to TagsComponent and Tags.  The state has no
effect yet.

/cc @adriancole

songy23 commented on core/src/main/java/io/opencensus/tags/TaggingState.java:
```diff
@@ -0,0 +1,39 @@
+   * <p>The {@link TagsComponent} may not add tags to {@link TagContext}s, propagate {@code
+   * TagContext}s in the current context, or serialize {@code TagContext}s.
+   */
+  DISABLED
```
nit: Will this state disable stats collection? If we plan to do so maybe worth mentioning in the comment.

sebright replied:
This is still under discussion in #608 (https://github.com/census-instrumentation/opencensus-java/issues/608#issuecomment-329828347), so I added a TODO.

bogdandrutu commented on core/src/main/java/io/opencensus/tags/TaggingState.java:
```diff
@@ -0,0 +1,40 @@
+package io.opencensus.tags;
+
+/** State of the {@link TagsComponent}. */
+public enum TaggingState {
```
I think it is clear if you add a third state NO_OP and describe that as a state which cannot be modified and cannot get into that state.

What do you think?

sebright replied:
You're right, users would probably want to know whether calling `setState(TaggingState.ENABLED)` will have an effect.

sebright replied:
You're right, users would probably want to know whether calling `setState(TaggingState.ENABLED)` would have an effect.  I'm not sure how we should represent it though.  The third state would be a valid output for getState, but not a valid input for setState.

adriancole replied:
I think we should just ignore on noop maybe linking to where noop is controlled. noop state shouldn't cause damage even to callers of this.

sebright replied:
> I think we should just ignore on noop maybe linking to where noop is controlled. noop state shouldn't cause damage even to callers of this.

That makes sense, especially since we will also ignore setting ENABLED/DISABLED when the state is NO_OP.

We could also have a separate method for checking whether the tagging implementation is no-op.  Then TaggingState could just represent the different levels of tagging functionality.

sebright replied:
I added a method `boolean isTaggingAvailable()` to Tags and TagsComponent.  I tried both options, but I thought that it was easier to document the new method than a new enum constant.  What do you think of this design?

bogdandrutu replied:
I personally think that it is wrong to have conceptually one state split into actually 2 APIs "is available" and "set/get state".

I think it is easier to have 3 states as I explained in a previous comment and explain possible transitions that can happen (as a state machine).

Why do you think it was hard to explain? I would explain this way:
* NO_IMPLEMENTATION. This is a permanent state that can be change only by recompiling the library and include the implementation.
* ENABLE. This is the default state of the library and transition from ENABLE to DISABLE can be made via setState. (also include what enable means, propagate tags, etc.)
* DISABLE. same as ENABLE

What do you think? Also we may make SetState throw a checked exception that force people to catch it and take the correct action (it will not be used that often, I don't expect to have more than 1 call to the SetState per binary/library).

bogdandrutu replied:
Ok. Let's do this way. Can you add the same changes to tracing library later?

sebright replied:
Sure, I opened an issue for the change: #673

bogdandrutu commented on core_impl/src/main/java/io/opencensus/implcore/tags/TagsComponentImplBase.java:
```diff
@@ -35,4 +41,15 @@ public Tagger getTagger() {
+
+  @Override
+  public void setState(TaggingState newState) {
+    state.set(checkNotNull(newState));
```
it is better to use checkNotNull(newState, "newState") because the message will tell you what variable was null. gRPC people told me to do that because will help with debugging.

sebright replied:
Done

bogdandrutu commented on core/src/main/java/io/opencensus/tags/Tags.java:
```diff
@@ -49,6 +49,35 @@ public static TagPropagationComponent getTagPropagationComponent() {
+   *
+   * @return {@code true} if a tagging implementation is available.
+   */
+  public static boolean isTaggingAvailable() {
```
Maybe isImplementationAvailable? Refer also to the readme how to do that with a @plainlink.

bogdandrutu commented on core/src/main/java/io/opencensus/tags/Tags.java:
```diff
@@ -49,6 +49,35 @@ public static TagPropagationComponent getTagPropagationComponent() {
+   *
+   * @param state the new {@code TaggingState}.
+   */
+  public static void setState(TaggingState state) {
```
Please comment that this is no-op if {@link #isImplementationAvailable()} returns false.

sebright replied:
Done

bogdandrutu commented on core/src/main/java/io/opencensus/tags/Tags.java:
```diff
@@ -49,6 +49,35 @@ public static TagPropagationComponent getTagPropagationComponent() {
+   *
+   * @return the current {@code TaggingState}.
+   */
+  public static TaggingState getState() {
```
Document that always returns DISABLE if {@link #isImplementationAvailable()} returns false.

sebright replied:
Done

adriancole commented:
Main thing is "when in doub leave it out" is a pretty good policy. Better to have users scream for something than have the code scream ;) In brave we kept it simple.. isNoop means dont bother.. doesnt matter why. We can do other things but I havent heard a request like this. I can check in spring if anyone else has encountered the need to split states.. if theres a workaround (like type check) we can at least defer the no impl state

All this said.. if strong feelings I can be on board with it.. so make a call!

bogdandrutu commented:
@adriancole not sure I understand what exactly do you propose here. Can you clarify?

Do you support disable at runtime? What is the default state?

adriancole commented:
Here's the background.

Request https://github.com/openzipkin/brave/issues/435
Impl https://github.com/openzipkin/brave/pull/445

Luckily we only had to make a simple switch to satisfy the customer in this
case, and we had no further requests. That doesnt mean more sophisticated
requests wont come in the future.

Still not sure if the feature matches what we are discussing but yeah hope
it helps.

adriancole commented:
On what is default state.. it is assumed tracing is enabled by default.
Because brave doesnt encourage static factory like Tracer.getDefault if you
have a tracer it is enabled (though if sampling is 0%  results are always
noop spans). We do have a noop span which is the result of a sampled false
decision. Brave 4 does up front sampling and doesnt retain data if
unsampled (noop span is same as unsampled span). Brave 5 probably will give
option to always record. Forgive if too much context.. not sure what you
want to know.

sebright commented:
@adriancole Thanks for the background.  That looks very similar to #608, including the comment about whether to disable propagation.  I didn't consider the need to stop stats/tracing functionality after startup, but I think this design will also support that use case.

sebright commented:
It sounds like we agree on at least providing get/setState methods, but not isImplementationAvailable().  How about I just remove the commits that added isImplementationAvailable(), and then we can decide whether we need to treat no-op differently in a separate issue.  @adriancole were you suggesting that we only use boolean, or do you think that the enum is okay?  I wanted to use an enum, because tracing will have three states (#608), and it will be easier to add states to tagging in the future.

bogdandrutu commented:
Sounds also good. We may not have 3 states for tracing at the beginning but I still think is better to have an extensible API.

You can remove that and mention in the comment that if no implementation linked the library will always return DISABLE and setState will be a no-op.

sebright commented:
I removed `isImplementationAvailable` and updated the Javadocs for `getState` and `setState`.  I also realized that we already have an issue for checking whether classes are no-op, #585, so we can continue the discussion there.

sebright commented:
@bogdandrutu @adriancole Do you mind if I merge this PR?  It only handles the enabled/disabled states, and we can discuss handling no-op in #585.

bogdandrutu commented:
@sebright go ahead.

sebright commented:
Thanks!

adriancole commented:
+1
